File: OpenStreetMap/iOS/FpsLabel.m
//
//  FpsLabel.m
//  Go Map!!
//
//  Created by Bryce Cogswell on 6/15/14.
//  Copyright (c) 2014 Bryce Cogswell. All rights reserved.
//

#import "DisplayLink.h"
#import "FpsLabel.h"

#define ENABLE_FPS	0


const int FRAME_COUNT = 60;

@implementation FpsLabel
{
	int					_historyPos;
	CFTimeInterval		_history[ FRAME_COUNT ];	// average last 60 frames
	dispatch_source_t	_timer;
}

-(void)awakeFromNib
{
	[super awakeFromNib];
	self.showFPS = NO;
	self.hidden = YES;
}

- (void)dealloc
{
	self.showFPS = NO;
}

-(void)setShowFPS:(BOOL)showFPS
{
	if ( showFPS != _showFPS ) {
		_showFPS = showFPS;

		if ( showFPS ) {

			self.hidden = NO;
			
			DisplayLink * displayLink = [DisplayLink shared];
			__weak FpsLabel * weakSelf = self;
			[displayLink addName:@"FpsLabel" block:^{
				[weakSelf frameUpdated];
			}];

			// create a timer to update the text twice a second
			dispatch_queue_t queue = dispatch_get_main_queue();
			_timer = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0, queue);
			if ( _timer ) {
				dispatch_source_set_timer(_timer, DISPATCH_TIME_NOW, NSEC_PER_SEC/2, NSEC_PER_SEC/5);
				dispatch_source_set_event_handler(_timer, ^{
					[self updateText];
				} );
				dispatch_resume(_timer);
			}
			self.layer.backgroundColor	= [UIColor colorWithWhite:1.0 alpha:0.6].CGColor;

		} else {
			self.text = nil;
			self.hidden = YES;
			if ( _timer ) {
				DisplayLink * displayLink = [DisplayLink shared];
				[displayLink removeName:@"FpsLabel"];
				dispatch_source_cancel( _timer );
			}
		}
	}
}


- (void)updateText
{
	// scan backward to see how many frames were drawn in the last second
	int frameCount = 0;
	int pos = (_historyPos + FRAME_COUNT - 1) % FRAME_COUNT;
	CFTimeInterval last = _history[ pos ];
	CFTimeInterval prev = 0.0;
	do {
		if ( --pos < 0 )
			pos = FRAME_COUNT - 1;
		prev = _history[pos];
		++frameCount;
		if ( last - prev >= 1.0 )
			break;
	} while ( pos != _historyPos );

	CFTimeInterval average = frameCount / (last - prev);
	if ( average >= 10.0 )
		self.text = [NSString stringWithFormat:@"%.1f FPS", average];
	else
		self.text = [NSString stringWithFormat:@"%.2f FPS", average];
	NSLog(@"frame");
}

- (void)frameUpdated
{
	// add to history
	CFTimeInterval now = CACurrentMediaTime();
	_history[_historyPos++] = now;
	if ( _historyPos >= FRAME_COUNT )
		_historyPos = 0;
}

@end

File: src/iOS/CustomViews/FpsLabel.swift
//
//  FpsLabel.swift
//  Go Map!!
//
//  Created by Bryce Cogswell on 6/15/14.
//  Copyright (c) 2014 Bryce Cogswell. All rights reserved.
//

import UIKit

private let ENABLE_FPS = 0

private let FRAME_COUNT = 2 * 60

final class FpsLabel: UILabel {
	private var historyPos = 0
	private var frameTimestamp = [CFTimeInterval](repeating: 0, count: FRAME_COUNT) // average last 60 frames

	private var timer: DispatchSourceTimer!

	public var showFPS = false {
		didSet {
			if showFPS == oldValue {
				return
			}
			if showFPS {
				isHidden = false
				DisplayLink.shared.addName("FpsLabel", block: {
					self.frameUpdated()
				})

				// create a timer to update the text twice a second
				timer = DispatchSource.makeTimerSource(queue: DispatchQueue.main)
				timer.schedule(deadline: .now(), repeating: .milliseconds(500))
				timer.setEventHandler(handler: { [weak self] in
					self?.updateText()
				})
				timer?.activate()
				layer.backgroundColor = UIColor(white: 1.0, alpha: 0.6).cgColor
			} else {
				text = nil
				isHidden = true
				DisplayLink.shared.removeName("FpsLabel")
				timer?.cancel()
			}
		}
	}

	override func awakeFromNib() {
		super.awakeFromNib()
		showFPS = false
		isHidden = true
	}

	deinit {
		showFPS = false
	}

	func updateText() {
		// scan backward to see how many frames were drawn in the last second
		let seconds = Double(FRAME_COUNT / 60)
		var frameCount: Double = 0
		var pos = (historyPos + FRAME_COUNT - 1) % FRAME_COUNT
		let last = frameTimestamp[pos]
		var prev: CFTimeInterval = 0.0
		repeat {
			pos -= 1
			if pos < 0 {
				pos = FRAME_COUNT - 1
			}
			prev = frameTimestamp[pos]
			frameCount += 1
			if last - prev >= seconds {
				break
			}
		} while pos != historyPos

		let average = CFTimeInterval(frameCount / (last - prev))
		if average >= 10.0 {
			text = String(format: "%.1f FPS", average)
		} else {
			text = String(format: "%.2f FPS", average)
		}
	}

	func frameUpdated() {
		// add to history
		let now = CACurrentMediaTime()
		frameTimestamp[historyPos] = now
		historyPos += 1
		if historyPos >= FRAME_COUNT {
			historyPos = 0
		}
	}
}

private let AUTOSCROLL_DISPLAYLINK_NAME = "autoScroll"

extension FpsLabel {

	var automatedFramerateTestActive: Bool {
		get {
			return DisplayLink.shared.hasName(AUTOSCROLL_DISPLAYLINK_NAME)
		}
		set(enable) {
			let displayLink = DisplayLink.shared
			let mainView = AppDelegate.shared.mainView!

			if enable == displayLink.hasName(AUTOSCROLL_DISPLAYLINK_NAME) {
				// unchanged
				return
			}

			if enable {
				// automaatically scroll view for frame rate testing
				showFPS = true

				// this set's the starting center point
				let startLatLon = LatLon(lon: -122.2060122462481, lat: 47.675389766549706)
				let startZoom = 18.0

				// sets the size of the circle
				let mpd = MetersPerDegreeAt(latitude: startLatLon.lat)
				let radius = 35.0
				let radius2 = CGPoint(x: radius / mpd.x, y: radius / mpd.y)
				let startTime = CACurrentMediaTime()
				let periodSeconds = 2.0

				displayLink.addName(AUTOSCROLL_DISPLAYLINK_NAME, block: {
					let offset = 1.0 - fmod((CACurrentMediaTime() - startTime) / periodSeconds, 1.0)
					let origin = LatLon(lon: startLatLon.lon + cos(offset * 2.0 * .pi) * radius2.x,
					                    lat: startLatLon.lat + sin(offset * 2.0 * .pi) * radius2.y)
					let zoomFrac = (1.0 + cos(offset * 2.0 * .pi)) * 0.5
					let zoom = startZoom * (1 + zoomFrac * 0.01)
					mainView.viewPort.centerOn(latLon: origin, zoom: zoom, rotation: nil)
				})
			} else {
				showFPS = false
				displayLink.removeName(AUTOSCROLL_DISPLAYLINK_NAME)
			}
		}
	}
}
